### ISSUE TITLE
학습 데이터셋 소개

### ISSUE THREAD

**Issue by chaeyoung-lee** [chaeyoung-lee/upstage-math-ocr#5]
# 수학수식 OCR 데이터셋

저자: 이채영 (cylee7133@gmail.com), UpstageAI

## 목차
1. [데이터셋 종류](https://github.com/chaeyoung-lee/upstage-math-ocr/wiki/%ED%95%99%EC%8A%B5-%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%85%8B#1-%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%85%8B-%EC%A2%85%EB%A5%98)
2. [전처리](https://github.com/chaeyoung-lee/upstage-math-ocr/wiki/%ED%95%99%EC%8A%B5-%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%85%8B#2-%EC%A0%84%EC%B2%98%EB%A6%AC)
3. [베이스라인 성능](https://github.com/chaeyoung-lee/upstage-math-ocr/wiki/%ED%95%99%EC%8A%B5-%EB%8D%B0%EC%9D%B4%ED%84%B0%EC%85%8B#3-%EB%B2%A0%EC%9D%B4%EC%8A%A4%EB%9D%BC%EC%9D%B8-%EC%84%B1%EB%8A%A5)
4. [Future Work]()

<br>
<br>

## 1. 데이터셋 종류

<br>

### 공개

- [Aida Calculus Math Handwriting Recognition Dataset (Aida)](https://www.kaggle.com/aidapearson/ocr-data)
- [Competition on Recognition of Online Handwritten Mathematical Expressions (CROHME)](https://www.isical.ac.in/~crohme/)
- [IM2LATEX-100k](https://zenodo.org/record/56198#.V2p0KTXT6eA)

### 비공개

- Upstage Math OCR Dataset

<br>

### Overview

| Dataset 	| # Formula 	| # Classes 	| Image Type 	| GT Type 	| License 	|
|:-:	|:-:	|:-:	|:-:	|:-:	|:-:	|
| Aida 	| 100000 	| 54 	| Synthetic handwritten 	| Character RoI, LaTeX, Mask 	| [CDLA-Sharing-1.0](https://cdla.dev/sharing-1-0/) 	|
| CROHME 	| 11637 	| 133 	| Online handwritten 	| LaTeX 	| [ CC BY-NC-SA 3.0](https://creativecommons.org/licenses/by-nc-sa/3.0/deed.en) 	|
| IM2LATEX PDF 	| 33986 	| 131 	| PDF (ArXiv) 	| LaTeX 	| [CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/legalcode) 	|
| IM2LATEX Handwritten 	| 41249 	| 147 	| Synthetic handwritten 	| LaTeX 	| [CC0 1.0 Universal](https://creativecommons.org/publicdomain/zero/1.0/legalcode) 	|
| Upstage Print 	| 82314 	| 212 	| Real print 	| Formula RoI, LaTeX 	| N/A 	|
| Upstage Handwritten 	| 55337 	| 195 	| Real handwritten 	| Formula RoI, LaTeX 	| N/A 	|

<br>
<br>

## 2. 전처리

<br>

### 결과물 다운로드

모든 데이터는 인식기 train-ready 포맷으로 이진원님과 문영기님 서버에 임시로 저장되어 있습니다. (Updated 05/31/2021: [GDrive](https://drive.google.com/drive/folders/1F3Rtw0xh4zpPKHGCKIAMYzDDjOeLFhgM?usp=sharing)에서 다운로드 가능)
- 이진원님 서버 내 위치: `/mnt/ssd/data`
- 문영기님 서버 내 위치: `/home/yk/Work/cylee/data`

두 서버 모두에서 `data` 폴더에는 `Aida, CROHME, IM2LATEX/PDF, IM2LATEX/Handwritten, Upstage` 폴더가 있고 각각의 데이터셋 폴더에는 다음과 같은 구조로 전처리가 완료된 데이터가 저장되어 있습니다.
```
[dataset]/
├── gt.txt
├── tokens.txt
└── images/
    ├── *.jpg
    ├── ...     
    └── *.jpg
```

또한 각각의 텍스트 파일은 다음과 같은 내용으로 작성되어 있습니다.
```
# inside gt.txt
[image name]\t[formula gt]
1_1.jpg\tx + y = 3
1_2.jpg\t\frac { 2 } { 3 } + 3 =
...

# inside tokens.txt
[token]
\sqrt
\frac
2
x
...
```

<br>

### 업스테이지 데이터 전처리 방법

업스테이지 데이터의 원본 폴더 또한 서버의 `data` 폴더에 `Upstage_raw`라는 이름으로 저장되어 있습니다. 원본 폴더는 대략 1000개의 이미지로 구성된 배치 폴더 여러개로 이루어져 있습니다. 각각의 배치 폴더 안에는 이미지 파일들과 GT 데이터가 저장된 하나의 JSON 파일이 있습니다. 각각의 이미지에는 다음과 같이 여러개의 수식이 존재하기 때문에 스탠더드한 수식인식기에 입력하기 위해선 수식 단위로 이미지를 잘라야 합니다.

![sample_upstage](https://user-images.githubusercontent.com/26811858/120123839-36c11c00-c1ec-11eb-9ead-5dd0d3e9032a.png)

이전 섹션에서 설명한 인식기 학습용 데이터 포맷으로의 전처리를 위해서는 다음과 같은 과정이 필요합니다.
- 배치별 JSON 파일에서 GT 정보 추출하기
- 이미지를 수식 단위로 자르기
- naming rule로 통일하기
- 클래스 맵 생성하기

처음 세가지 태스크는 본 리포의 `src/data_tools/parse_upstage.py`를 이용합니다.
```bash
python parse_upstage.py -i upstage_raw/ -o processed/
```
이때 `-academic_level` 플래그를 추가하면 파싱과 동시에 JSON 파일에서 수식별 난이도 데이터를 추출해 `output_folder/academic_level.txt` 파일에 저장합니다.

마지막 클래스 맵 생성은 `src/data_tools/extract_tokens.py`를 이용합니다.
```bash
python extract_tokens.py -o processed/tokens.txt processed/gt.txt
```
이렇게 생성된 클래스 맵은 이후 인식기를 학습할 때 `gt.txt`와 함께 인풋으로 이용됩니다.

<br>
<br>

## 3. 베이스라인 성능

본 리포는 두가지 모델을 지원하고 있습니다.
- Attention
- SATRN

벤치마크 데이터셋과 업스테이지 데이터셋 각각에서의 두 모델을 학습했습니다. 모든 학습에서의 컴퓨팅 환경은 다음과 같습니다.

- PyTorch 1.4.0
- CUDA 10.0, cuDNN 7.6.3
- Tesla V100 32G

<br>

### Attention Model

- Encoder: CNN (5.5M)
- Decoder: LSTM with Attention (0.5M)
- Loss: Cross Entropy
- Optimizer: Adam
- Epoch: 50
- Training Time: (epoch 당 ~10분)

| Dataset 	| Train # 	| Test # 	| Symbol Accuracy 	| Sentence Accuracy 	| Word Error Rate 	|
|:-:	|:-:	|:-:	|:-:	|:-:	|:-:	|
| Aida 	|  80000	| 20000 	|  94.46	|  	|  	|
| CROHME 	|  	|  	|  	|  	|  	|
| IM2LATEX PDF 	|  	|  	|  	|  	|  	|
| IM2LATEX Handwritten 	|  	|  	|  	|  	|  	|
| Upstage Print 	|  	|  	|  71.91	|  	|  	|
| Upstage Handwritten 	|  	|  	|  	|  	|  	|

<br>

### SATRN Model

- Encoder: CNN + Transformer (5.8M)
- Decoder: Transformer (0.9M)
- Loss: Cross Entropy
- Optimizer: Adam
- Epoch: 50
- Training Time: (epoch 당 ~30분)

| Dataset 	| Train # 	| Test # 	| Symbol Accuracy 	| Sentence Accuracy 	| Word Error Rate 	|
|:-:	|:-:	|:-:	|:-:	|:-:	|:-:	|
| Aida 	|  80000	| 20000 	|  	|  	|  	|
| CROHME 	|  	|  	|  	|  	|  	|
| IM2LATEX  | 60190 	| 15047 	|  	|  	|  	|
| Upstage Print+HR |  80000|  20000|  	76.00|  69.11|  0.11|

<br>

### Checkpoints

모든 모델 체크포인트는 마찬가지로 이진원님과 문영기님 서버에 임시로 저장되어 있습니다.
- 이진원님 서버 내 위치: `/mnt/ssd/checkpoints`
- 문영기님 서버 내 위치: `/home/yk/Work/cylee/checkpoints`

`checkpoints` 폴더에는 각각의 모델 파일이 `[model]_[dataset]_[epoch].pth` 과 같은 형태로 저장되어 있습니다. 또한 같은 위치에 존재하는 `[model]_[dataset]_[epoch].yaml` 를 통해 각 모델의 자세한 학습 세팅을 볼 수 있습니다.

<br>

### RECT -> QUAD

업스테이지 원본 데이터의 수식 Bounding Box는 QUAD (x1, x2, x3, x4, y1, y2, y3, y4) 형태로 라벨링됐습니다. 하지만 데이터 전처리의 편의성, 유스테이지 강의 대회의 난이도, 모델 robustness를 위해서 현재 전처리된 버젼의 데이터는 RECT (x, y, w, h) 형태로 수식별 Bounding Box를 잘랐습니다.

### Rotation

Aspect ratio(width/height)가 1 미만인 이미지를 모았습니다.

### User Report

유스테이지 이진원님의 수식인식 강의에서 본 데이터셋을 대회용으로 활용한 결과 강의 학생들에게 데이터셋 이슈 관련해서 받은 유저 리포트입니다.
